docs/adr/0003-delivery-channels-templating.md

```diff
@@ -0,0 +1,92 @@
+# Templating engine for different notification types
+
+## Status
+
+Proposed
+
+## Context
+
+This ADR proposes a method on how to implement custom templating when a certain notification is sent through a communication channels (SMS, email, web push notifications etc).
+
+In a typical notification system, an object will be applied to a basic template to provide an end user with information targeting them. A lot of the basic design/layout of the template will be quite standard, but there will be places for values that will need to be inserted to customise it for the said end user. An example of this would be an birthday email notification, where a lot of the design and formatting of the email/subject will already be very well defined, and then insertion of custom data (eg. email id, name, etc ) is what will be needed by a templater.
+
+The content of this ADR will work off themes from the previous ADR (0002-delivery-channel-creation-and-integration.md).
+
+## Decision
+
+A notification having necessary information needs to be translated into a certain communication object (SMS, Email, etc) before it can be sent via the desired delivery channel (SendGrid, Twillio, Airship).
+
+An example of this would be a SMS, which needs the essential information to be available for it to be converted into an SMS object before being sent to the end user.
+
+```json
+
+    public class SMS {
+        public string ToPhoneNumber { get; set; }
+        public string Body { get; set; }
+    }
+
+```
+
+Each field in the communication object will needs it's own template (property) for the associated notification type. A template for the `ToPhoneNumber` property in the above example would look something like:
+
+```
+{{ notification.propertyBag.ToPhoneNumber }}
+```
+
+### Creation of templates, how does someone create templates and in what language
+
+A template will be created based on the notification type, the communication type, and the communication object's property key.
```

Comment:
It looks from this section as templates will always need to be supplied in sets. E.g., if you provide a `(New Lead, Email, ToEmail)` template, this is of no use unless you also supply `(New Lead, Email, Subject)` and `(New Lead, Email, Body)` templates.

It would probably be worth saying that explicitly here.

This in turn implies you'll need to do one of two things. Either a) require all the necessary templates for a particular `(New Lead, Email, xx)` combination to be registered at once or b) define what you will do if an attempt is made to send a `(Notification Type, Communication Type)` combination in cases there are missing templates.

Reply:
That is correct. If the Email was to be sent, then it means that we would need all the templates necessary and the user settings stored in the user preferences table otherwise an email could not be triggered.

Reply:
OK, so let's make that explicit in the ADR please.